File: src/sokoban.ts
import {
  type HardVerifierResult,
  type ProposalTrace,
  type Receipt,
  type ReplayRollbackAdapter,
  type TypedCandidate,
  Ledger,
  TransactionEngine,
  hardAccept,
  hardReject,
  makeCandidate,
  makeTrace,
} from "./core.js";
import { stableHash } from "./canonical.js";

export type SokobanPosition = [number, number];
export type SokobanDirection = "U" | "D" | "L" | "R";

export interface SokobanLayout {
  height: number;
  width: number;
  walls: SokobanPosition[];
  goals: SokobanPosition[];
}

export interface SokobanState {
  boxes: SokobanPosition[];
  player: SokobanPosition;
}

export interface SokobanPush {
  box: SokobanPosition;
  direction: SokobanDirection;
}

export interface SokobanCandidatePayload {
  layout: SokobanLayout;
  solvedState: SokobanState;
  predecessor: SokobanState;
  pushes: SokobanPush[];
  cost: number;
}

export interface SokobanReverseReport {
  solved: boolean;
  predecessor: SokobanState | null;
  solvedState: SokobanState;
  pushes: SokobanPush[];
  verifierCalls: number;
  reverseExpansions: number;
  maxDepth: number;
  baselineStateCount: number;
  verifierCallReduction: number;
  invalidCommitCount: number;
  ledgerAudit: boolean;
  replayRollbackRate: number;
}

const DIRECTIONS: Array<[SokobanDirection, SokobanPosition]> = [
  ["U", [-1, 0]],
  ["D", [1, 0]],
  ["L", [0, -1]],
  ["R", [0, 1]],
];

interface SearchNode {
  state: SokobanState;
  pushes: SokobanPush[];
  depth: number;
}

export function parseSokoban(rows: string[]): { layout: SokobanLayout; state: SokobanState } {
  if (rows.length === 0) {
    throw new RangeError("sokoban level must not be empty");
  }
  const width = rows[0].length;
  if (width === 0 || rows.some((row) => row.length !== width)) {
    throw new RangeError("sokoban rows must be non-empty and equal width");
  }
  const walls: SokobanPosition[] = [];
  const goals: SokobanPosition[] = [];
  const boxes: SokobanPosition[] = [];
  const players: SokobanPosition[] = [];
  const valid = new Set(["#", " ", ".", "$", "*", "@", "+"]);
  for (let row = 0; row < rows.length; row += 1) {
    for (let col = 0; col < width; col += 1) {
      const cell = rows[row][col];
      if (!valid.has(cell)) {
        throw new RangeError(`unsupported sokoban tile: ${cell}`);
      }
      const pos: SokobanPosition = [row, col];
      if (cell === "#") walls.push(pos);
      if (cell === "." || cell === "*" || cell === "+") goals.push(pos);
      if (cell === "$" || cell === "*") boxes.push(pos);
      if (cell === "@" || cell === "+") players.push(pos);
    }
  }
  if (players.length !== 1) {
    throw new RangeError("sokoban level must contain exactly one player");
  }
  if (goals.length === 0) {
    throw new RangeError("sokoban level must contain at least one goal");
  }
  if (boxes.length !== goals.length) {
    throw new RangeError("sokoban level must contain the same number of boxes and goals");
  }
  return {
    layout: normalizeLayout({ height: rows.length, width, walls, goals }),
    state: normalizeState({ boxes, player: players[0] }),
  };
}

export function renderSokoban(layoutInput: SokobanLayout, stateInput: SokobanState): string[] {
  const layout = normalizeLayout(layoutInput);
  const state = normalizeState(stateInput);
  const walls = positionSet(layout.walls);
  const goals = positionSet(layout.goals);
  const boxes = positionSet(state.boxes);
  const rows: string[] = [];
  for (let row = 0; row < layout.height; row += 1) {
    let out = "";
    for (let col = 0; col < layout.width; col += 1) {
      const key = positionKey([row, col]);
      if (walls.has(key)) out += "#";
      else if (samePosition(state.player, [row, col]) && goals.has(key)) out += "+";
      else if (samePosition(state.player, [row, col])) out += "@";
      else if (boxes.has(key) && goals.has(key)) out += "*";
      else if (boxes.has(key)) out += "$";
      else if (goals.has(key)) out += ".";
      else out += " ";
    }
    rows.push(out);
  }
  return rows;
}

export function sokobanSolved(layout: SokobanLayout, state: SokobanState): boolean {
  const goals = positionSet(normalizeLayout(layout).goals);
  const boxes = positionSet(normalizeState(state).boxes);
  return boxes.size === goals.size && Array.from(boxes).every((box) => goals.has(box));
}

export async function makeSokobanCandidate(
  layoutInput: SokobanLayout,
  solvedStateInput: SokobanState,
  predecessorInput: SokobanState,
  pushesInput: SokobanPush[],
  cost: number,
): Promise<TypedCandidate<SokobanCandidatePayload>> {
  const layout = normalizeLayout(layoutInput);
  const solvedState = normalizeState(solvedStateInput);
  const predecessor = normalizeState(predecessorInput);
  const pushes = pushesInput.map(normalizePush);
  return makeCandidate(
    { layout, solvedState, predecessor, pushes, cost },
    "sokoban.reverse_certificate",
    "sokoban.reverse_certificate.v1",
    {
      layout: await stableHash(layout),
      solvedState: await stableHash(solvedState),
      predecessor: await stableHash(predecessor),
      pushes: await stableHash(pushes),
    },
  );
}

export class SokobanReverseAdapter implements ReplayRollbackAdapter<SokobanState, SokobanCandidatePayload> {
  verifierId = "sokoban_forward_replay_verifier";
  verifierVersion = "1.0";

  verify(candidate: TypedCandidate<SokobanCandidatePayload>): HardVerifierResult {
    const payload = candidate.payload;
    const layout = normalizeLayout(payload.layout);
    const solvedState = normalizeState(payload.solvedState);
    const predecessor = normalizeState(payload.predecessor);
    const pushes = payload.pushes.map(normalizePush);
    const metadata = { cost: payload.cost, pushCount: pushes.length };
    try {
      const replayed = replaySokobanPushes(layout, predecessor, pushes);
      if (statesEqual(replayed, solvedState) && sokobanSolved(layout, replayed)) {
        return hardAccept(this.verifierId, this.verifierVersion, metadata);
      }
      return hardReject(
        this.verifierId,
        this.verifierVersion,
        { replayed, solvedState, replayedSolved: sokobanSolved(layout, replayed) },
        metadata,
      );
    } catch (error) {
      return hardReject(
        this.verifierId,
        this.verifierVersion,
        { reason: error instanceof Error ? error.message : "unknown", predecessor },
        metadata,
      );
    }
  }

  applyCommit(_state: SokobanState, candidate: TypedCandidate<SokobanCandidatePayload>): SokobanState {
    return normalizeState(candidate.payload.predecessor);
  }

  replay(_state: SokobanState, receipt: Receipt): SokobanState {
    const payload = (receipt.replayBundle as { candidatePayload: SokobanCandidatePayload }).candidatePayload;
    return normalizeState(payload.predecessor);
  }

  rollback(_state: SokobanState, receipt: Receipt): SokobanState {
    return normalizeState((receipt.rollbackBundle as { preState: SokobanState }).preState);
  }
}

export function replaySokobanPushes(layout: SokobanLayout, state: SokobanState, pushes: SokobanPush[]): SokobanState {
  return pushes.reduce((current, push) => applySokobanPush(layout, current, push), normalizeState(state));
}

export function applySokobanPush(layoutInput: SokobanLayout, stateInput: SokobanState, pushInput: SokobanPush): SokobanState {
  const layout = normalizeLayout(layoutInput);
  const state = normalizeState(stateInput);
  const push = normalizePush(pushInput);
  const boxes = new Map(state.boxes.map((box) => [positionKey(box), box]));
  const boxKey = positionKey(push.box);
  if (!boxes.has(boxKey)) {
    throw new Error(`no box at push origin: ${boxKey}`);
  }
  const delta = directionDelta(push.direction);
  const stand = subPosition(push.box, delta);
  const destination = addPosition(push.box, delta);
  if (!isOpen(layout, state.boxes, stand)) {
    throw new Error(`push stand cell is blocked: ${positionKey(stand)}`);
  }
  if (!reachable(layout, state.boxes, state.player, stand)) {
    throw new Error(`player cannot reach push stand cell: ${positionKey(stand)}`);
  }
  boxes.delete(boxKey);
  const remaining = Array.from(boxes.values());
  if (!isOpen(layout, remaining, destination)) {
    throw new Error(`push destination is blocked: ${positionKey(destination)}`);
  }
  remaining.push(destination);
  return normalizeState({ boxes: remaining, player: push.box });
}

export function reachable(layoutInput: SokobanLayout, boxesInput: SokobanPosition[], start: SokobanPosition, target: SokobanPosition): boolean {
  const layout = normalizeLayout(layoutInput);
  const boxes = positionSet(boxesInput);
  const startKey = positionKey(start);
  const targetKey = positionKey(target);
  if (boxes.has(startKey) || boxes.has(targetKey) || !inBounds(layout, start) || !inBounds(layout, target)) {
    return false;
  }
  const walls = positionSet(layout.walls);
  if (walls.has(startKey) || walls.has(targetKey)) {
    return false;
  }
  const queue: SokobanPosition[] = [start];
  const seen = new Set([startKey]);
  for (let idx = 0; idx < queue.length; idx += 1) {
    const current = queue[idx];
    if (samePosition(current, target)) {
      return true;
    }
    for (const [, delta] of DIRECTIONS) {
      const next = addPosition(current, delta);
      const key = positionKey(next);
      if (seen.has(key) || boxes.has(key) || walls.has(key) || !inBounds(layout, next)) {
        continue;
      }
      seen.add(key);
      queue.push(next);
    }
  }
  return false;
}

export async function reversePullTraces(
  layoutInput: SokobanLayout,
  solvedStateInput: SokobanState,
  maxDepth = 3,
  maxCandidates = 64,
): Promise<Array<{ trace: ProposalTrace; candidate: TypedCandidate<SokobanCandidatePayload> }>> {
  return (await collectReversePullTraces(layoutInput, solvedStateInput, maxDepth, maxCandidates)).items;
}

async function collectReversePullTraces(
  layoutInput: SokobanLayout,
  solvedStateInput: SokobanState,
  maxDepth = 3,
  maxCandidates = 64,
): Promise<{ items: Array<{ trace: ProposalTrace; candidate: TypedCandidate<SokobanCandidatePayload> }>; expansions: number }> {
  if (!Number.isInteger(maxDepth) || maxDepth <= 0) {
    throw new RangeError("maxDepth must be positive");
  }
  if (!Number.isInteger(maxCandidates) || maxCandidates <= 0) {
    throw new RangeError("maxCandidates must be positive");
  }
  const layout = normalizeLayout(layoutInput);
  const solvedState = normalizeState(solvedStateInput);
  const queue: SearchNode[] = [{ state: solvedState, pushes: [], depth: 0 }];
  const seen = new Set([stateKey(solvedState)]);
  const candidates: Array<{ trace: ProposalTrace; candidate: TypedCandidate<SokobanCandidatePayload> }> = [];
  let expansions = 0;
  for (let queueIdx = 0; queueIdx < queue.length && candidates.length < maxCandidates; queueIdx += 1) {
    const node = queue[queueIdx];
    if (node.depth >= maxDepth) {
      continue;
    }
    expansions += 1;
    for (const box of node.state.boxes) {
      for (const [direction, delta] of DIRECTIONS) {
        const previousBox = subPosition(box, delta);
        const previousPlayer = subPosition(previousBox, delta);
        if (!reachable(layout, node.state.boxes, node.state.player, previousBox)) {
          continue;
        }
        const boxes = node.state.boxes.filter((candidateBox) => !samePosition(candidateBox, box));
        if (!isOpen(layout, boxes, previousBox)) {
          continue;
        }
        boxes.push(previousBox);
        const predecessor = normalizeState({ boxes, player: previousPlayer });
        if (!isOpen(layout, predecessor.boxes, previousPlayer)) {
          continue;
        }
        const key = stateKey(predecessor);
        if (seen.has(key)) {
          continue;
        }
        seen.add(key);
        const push: SokobanPush = { box: previousBox, direction };
        const pushes = [push, ...node.pushes];
        const depth = node.depth + 1;
        const trace = makeTrace({
          branchId: `sokoban-reverse-${candidates.length}`,
          actions: pushes,
          seeds: [await stableHash(solvedState), depth, candidates.length],
          modelVersion: "reverse_pull.sokoban.v1",
        });
        const candidate = await makeSokobanCandidate(layout, solvedState, predecessor, pushes, depth);
        candidates.push({ trace, candidate });
        queue.push({ state: predecessor, pushes, depth });
        if (candidates.length >= maxCandidates) {
          break;
        }
      }
      if (candidates.length >= maxCandidates) {
        break;
      }
    }
  }
  return { items: candidates, expansions };
}

export async function searchSokobanPredecessor(
  layoutInput: SokobanLayout,
  solvedStateInput: SokobanState,
  maxDepth = 3,
  maxCandidates = 64,
): Promise<SokobanReverseReport> {
  const layout = normalizeLayout(layoutInput);
  const solvedState = normalizeState(solvedStateInput);
  if (!sokobanSolved(layout, solvedState)) {
    throw new RangeError("solvedState must have every box on a goal");
  }
  const { items: traces, expansions } = await collectReversePullTraces(layout, solvedState, maxDepth, maxCandidates);
  const adapter = new SokobanReverseAdapter();
  const ledger = new Ledger();
  const engine = new TransactionEngine(adapter, ledger);
  let predecessor: SokobanState | null = null;
  let pushes: SokobanPush[] = [];
  for (const { trace, candidate } of traces) {
    const outcome = await engine.transact(solvedState, trace, candidate);
    if (outcome.committed) {
      predecessor = outcome.state;
      pushes = candidate.payload.pushes;
      break;
    }
  }
  const ledgerAudit = await ledger.audit();
  let replayRollbackRate = 0;
  if (ledgerAudit) {
    try {
      await engine.replayAudit(solvedState);
      const rollback = await engine.rollbackAudit(solvedState);
      replayRollbackRate = statesEqual(rollback, solvedState) ? 1 : 0;
    } catch (_error) {
      replayRollbackRate = 0;
    }
  }
  const baselineStateCount = baselineStateCountFor(layout);
  return {
    solved: predecessor !== null,
    predecessor,
    solvedState,
    pushes,
    verifierCalls: engine.hardVerifierCalls,
    reverseExpansions: expansions,
    maxDepth,
    baselineStateCount,
    verifierCallReduction: engine.hardVerifierCalls > 0 ? baselineStateCount / engine.hardVerifierCalls : 0,
    invalidCommitCount: engine.invalidCommitCount,
    ledgerAudit,
    replayRollbackRate,
  };
}

export function normalizeLayout(layout: SokobanLayout): SokobanLayout {
  if (!Number.isInteger(layout.height) || !Number.isInteger(layout.width) || layout.height <= 0 || layout.width <= 0) {
    throw new RangeError("sokoban layout dimensions must be positive integers");
  }
  return {
    height: layout.height,
    width: layout.width,
    walls: normalizePositions(layout.walls),
    goals: normalizePositions(layout.goals),
  };
}

export function normalizeState(state: SokobanState): SokobanState {
  return {
    boxes: normalizePositions(state.boxes),
    player: normalizePosition(state.player),
  };
}

function normalizePush(push: SokobanPush): SokobanPush {
  return {
    box: normalizePosition(push.box),
    direction: normalizeDirection(push.direction),
  };
}

function normalizePositions(positions: SokobanPosition[]): SokobanPosition[] {
  const normalized = positions.map(normalizePosition).sort(comparePositions);
  const unique = new Set(normalized.map(positionKey));
  if (unique.size !== normalized.length) {
    throw new RangeError("duplicate sokoban positions are not allowed");
  }
  return normalized;
}

function normalizePosition(position: SokobanPosition): SokobanPosition {
  if (!Array.isArray(position) || position.length !== 2 || !Number.isInteger(position[0]) || !Number.isInteger(position[1])) {
    throw new RangeError("sokoban positions must be integer pairs");
  }
  return [position[0], position[1]];
}

function normalizeDirection(direction: string): SokobanDirection {
  if (direction === "U" || direction === "D" || direction === "L" || direction === "R") {
    return direction;
  }
  throw new RangeError(`unknown sokoban direction: ${direction}`);
}

function directionDelta(direction: SokobanDirection): SokobanPosition {
  const found = DIRECTIONS.find(([name]) => name === direction);
  if (!found) {
    throw new RangeError(`unknown sokoban direction: ${direction}`);
  }
  return found[1];
}

function baselineStateCountFor(layout: SokobanLayout): number {
  const floor = layout.height * layout.width - layout.walls.length;
  return Math.max(0, floor * Math.max(0, floor - 1));
}

function isOpen(layout: SokobanLayout, boxes: SokobanPosition[], pos: SokobanPosition): boolean {
  return inBounds(layout, pos) && !positionSet(layout.walls).has(positionKey(pos)) && !positionSet(boxes).has(positionKey(pos));
}

function inBounds(layout: SokobanLayout, pos: SokobanPosition): boolean {
  return pos[0] >= 0 && pos[0] < layout.height && pos[1] >= 0 && pos[1] < layout.width;
}

function positionSet(positions: SokobanPosition[]): Set<string> {
  return new Set(positions.map(positionKey));
}

function positionKey(pos: SokobanPosition): string {
  return `${pos[0]},${pos[1]}`;
}

function stateKey(state: SokobanState): string {
  const normalized = normalizeState(state);
  return `${normalized.boxes.map(positionKey).join(";")}|${positionKey(normalized.player)}`;
}

function samePosition(a: SokobanPosition, b: SokobanPosition): boolean {
  return a[0] === b[0] && a[1] === b[1];
}

function statesEqual(a: SokobanState, b: SokobanState): boolean {
  return stateKey(a) === stateKey(b);
}

function addPosition(a: SokobanPosition, b: SokobanPosition): SokobanPosition {
  return [a[0] + b[0], a[1] + b[1]];
}

function subPosition(a: SokobanPosition, b: SokobanPosition): SokobanPosition {
  return [a[0] - b[0], a[1] - b[1]];
}

function comparePositions(a: SokobanPosition, b: SokobanPosition): number {
  if (a[0] !== b[0]) return a[0] - b[0];
  return a[1] - b[1];
}

File: trwm/experiments/sokoban.py
from __future__ import annotations

from collections import deque
from dataclasses import dataclass
from typing import Any, Iterable, Mapping

from ..core import HardVerifierResult, Ledger, ProposalTrace, TransactionEngine, TypedCandidate, stable_hash


Position = tuple[int, int]
PushAction = Mapping[str, Any]
DIRECTIONS: tuple[tuple[str, Position], ...] = (
    ("U", (-1, 0)),
    ("D", (1, 0)),
    ("L", (0, -1)),
    ("R", (0, 1)),
)


@dataclass(frozen=True)
class SokobanLayout:
    height: int
    width: int
    walls: tuple[Position, ...]
    goals: tuple[Position, ...]

    @property
    def wall_set(self) -> frozenset[Position]:
        return frozenset(self.walls)

    @property
    def goal_set(self) -> frozenset[Position]:
        return frozenset(self.goals)


@dataclass(frozen=True)
class SokobanState:
    boxes: tuple[Position, ...]
    player: Position


@dataclass(frozen=True)
class SokobanSearchNode:
    state: SokobanState
    pushes: tuple[PushAction, ...]
    depth: int


@dataclass(frozen=True)
class SokobanReverseReport:
    solved: bool
    predecessor: SokobanState | None
    solved_state: SokobanState
    pushes: tuple[PushAction, ...]
    verifier_calls: int
    reverse_expansions: int
    max_depth: int
    baseline_state_count: int
    verifier_call_reduction: float
    invalid_commit_count: int
    ledger_audit: bool
    replay_rollback_rate: float


def parse_sokoban(rows: Iterable[str]) -> tuple[SokobanLayout, SokobanState]:
    lines = tuple(str(row).rstrip("\n") for row in rows)
    if not lines:
        raise ValueError("sokoban level must not be empty")
    width = len(lines[0])
    if width == 0:
        raise ValueError("sokoban rows must not be empty")
    if any(len(line) != width for line in lines):
        raise ValueError("sokoban rows must have equal width")

    walls: list[Position] = []
    goals: list[Position] = []
    boxes: list[Position] = []
    players: list[Position] = []
    valid = {"#", " ", ".", "$", "*", "@", "+"}
    for row, line in enumerate(lines):
        for col, cell in enumerate(line):
            if cell not in valid:
                raise ValueError(f"unsupported sokoban tile: {cell!r}")
            pos = (row, col)
            if cell == "#":
                walls.append(pos)
            if cell in {".", "*", "+"}:
                goals.append(pos)
            if cell in {"$", "*"}:
                boxes.append(pos)
            if cell in {"@", "+"}:
                players.append(pos)
    if len(players) != 1:
        raise ValueError("sokoban level must contain exactly one player")
    if not goals:
        raise ValueError("sokoban level must contain at least one goal")
    if len(boxes) != len(goals):
        raise ValueError("sokoban level must contain the same number of boxes and goals")
    layout = SokobanLayout(
        height=len(lines),
        width=width,
        walls=tuple(sorted(walls)),
        goals=tuple(sorted(goals)),
    )
    return layout, SokobanState(boxes=_normalize_boxes(boxes), player=players[0])


def render_sokoban(layout: SokobanLayout, state: SokobanState) -> tuple[str, ...]:
    walls = layout.wall_set
    goals = layout.goal_set
    boxes = frozenset(state.boxes)
    rows: list[str] = []
    for row in range(layout.height):
        out = []
        for col in range(layout.width):
            pos = (row, col)
            if pos in walls:
                out.append("#")
            elif pos == state.player and pos in goals:
                out.append("+")
            elif pos == state.player:
                out.append("@")
            elif pos in boxes and pos in goals:
                out.append("*")
            elif pos in boxes:
                out.append("$")
            elif pos in goals:
                out.append(".")
            else:
                out.append(" ")
        rows.append("".join(out))
    return tuple(rows)


def is_solved(layout: SokobanLayout, state: SokobanState) -> bool:
    return frozenset(state.boxes) == layout.goal_set


def make_sokoban_candidate(
    layout: SokobanLayout,
    solved_state: SokobanState,
    predecessor: SokobanState,
    pushes: Iterable[PushAction],
    cost: int,
) -> TypedCandidate:
    pushes_tuple = tuple(_normalize_push(push) for push in pushes)
    return TypedCandidate(
        payload={
            "layout": layout,
            "solved_state": _normalize_state(solved_state),
            "predecessor": _normalize_state(predecessor),
            "pushes": pushes_tuple,
            "cost": int(cost),
        },
        type_name="sokoban.reverse_certificate",
        schema_version="sokoban.reverse_certificate.v1",
        hashes={
            "layout": stable_hash(layout),
            "solved_state": stable_hash(_normalize_state(solved_state)),
            "predecessor": stable_hash(_normalize_state(predecessor)),
            "pushes": stable_hash(pushes_tuple),
        },
    )


class SokobanReverseAdapter:
    verifier_id = "sokoban_forward_replay_verifier"
    verifier_version = "1.0"

    def verify(self, candidate: TypedCandidate) -> HardVerifierResult:
        payload = candidate.payload
        layout = _coerce_layout(payload["layout"])
        solved_state = _coerce_state(payload["solved_state"])
        predecessor = _coerce_state(payload["predecessor"])
        pushes = tuple(_normalize_push(push) for push in payload["pushes"])
        metadata = {"cost": payload.get("cost", len(pushes)), "push_count": len(pushes)}
        try:
            replayed = replay_pushes(layout, predecessor, pushes)
        except ValueError as exc:
            return HardVerifierResult.reject(
                self.verifier_id,
                self.verifier_version,
                residual={"reason": str(exc), "predecessor_hash": stable_hash(predecessor)},
                metadata=metadata,
            )
        if replayed == solved_state and is_solved(layout, replayed):
            return HardVerifierResult.accept(self.verifier_id, self.verifier_version, metadata=metadata)
        return HardVerifierResult.reject(
            self.verifier_id,
            self.verifier_version,
            residual={
                "replayed_hash": stable_hash(replayed),
                "solved_hash": stable_hash(solved_state),
                "replayed_solved": is_solved(layout, replayed),
            },
            metadata=metadata,
        )

    def apply_commit(self, state: SokobanState, candidate: TypedCandidate) -> SokobanState:
        return _coerce_state(candidate.payload["predecessor"])

    def replay(self, state: SokobanState, receipt: Any) -> SokobanState:
        return _coerce_state(receipt.replay_bundle["candidate_payload"]["predecessor"])

    def rollback(self, state: SokobanState, receipt: Any) -> SokobanState:
        pre_state = receipt.rollback_bundle["pre_state"]
        return _coerce_state(pre_state)


def replay_pushes(layout: SokobanLayout, state: SokobanState, pushes: Iterable[PushAction]) -> SokobanState:
    current = _normalize_state(state)
    for push in pushes:
        current = apply_push(layout, current, push)
    return current


def apply_push(layout: SokobanLayout, state: SokobanState, push: PushAction) -> SokobanState:
    action = _normalize_push(push)
    box = action["box"]
    delta = _direction_delta(action["direction"])
    boxes = set(state.boxes)
    if box not in boxes:
        raise ValueError(f"no box at push origin: {box}")
    stand = _sub(box, delta)
    dest = _add(box, delta)
    if not _is_open(layout, state.boxes, stand):
        raise ValueError(f"push stand cell is blocked: {stand}")
    if not reachable(layout, state.boxes, state.player, stand):
        raise ValueError(f"player cannot reach push stand cell: {stand}")
    boxes.remove(box)
    if not _is_open(layout, tuple(sorted(boxes)), dest):
        raise ValueError(f"push destination is blocked: {dest}")
    boxes.add(dest)
    return SokobanState(boxes=_normalize_boxes(boxes), player=box)


def reachable(layout: SokobanLayout, boxes: Iterable[Position], start: Position, target: Position) -> bool:
    box_set = frozenset(boxes)
    if start in box_set or target in box_set:
        return False
    if not _in_bounds(layout, start) or not _in_bounds(layout, target):
        return False
    if start in layout.wall_set or target in layout.wall_set:
        return False
    queue: deque[Position] = deque([start])
    seen = {start}
    while queue:
        current = queue.popleft()
        if current == target:
            return True
        for _, delta in DIRECTIONS:
            nxt = _add(current, delta)
            if nxt in seen or nxt in box_set or nxt in layout.wall_set or not _in_bounds(layout, nxt):
                continue
            seen.add(nxt)
            queue.append(nxt)
    return False


def reverse_pull_traces(
    layout: SokobanLayout,
    solved_state: SokobanState,
    max_depth: int = 3,
    max_candidates: int = 64,
) -> tuple[tuple[ProposalTrace, TypedCandidate], int]:
    if max_depth <= 0:
        raise ValueError("max_depth must be positive")
    if max_candidates <= 0:
        raise ValueError("max_candidates must be positive")
    solved_state = _normalize_state(solved_state)
    queue: deque[SokobanSearchNode] = deque([SokobanSearchNode(solved_state, (), 0)])
    seen = {_state_key(solved_state)}
    candidates: list[tuple[ProposalTrace, TypedCandidate]] = []
    expansions = 0

    while queue and len(candidates) < max_candidates:
        node = queue.popleft()
        if node.depth >= max_depth:
            continue
        expansions += 1
        for box in node.state.boxes:
            for direction, delta in DIRECTIONS:
                previous_box = _sub(box, delta)
                previous_player = _sub(previous_box, delta)
                if not reachable(layout, node.state.boxes, node.state.player, previous_box):
                    continue
                boxes = set(node.state.boxes)
                boxes.remove(box)
                if not _is_open(layout, tuple(sorted(boxes)), previous_box):
                    continue
                boxes.add(previous_box)
                predecessor = SokobanState(boxes=_normalize_boxes(boxes), player=previous_player)
                if not _is_open(layout, predecessor.boxes, previous_player):
                    continue
                key = _state_key(predecessor)
                if key in seen:
                    continue
                seen.add(key)
                push = {"box": previous_box, "direction": direction}
                pushes = (push,) + node.pushes
                depth = node.depth + 1
                trace = ProposalTrace(
                    branch_id=f"sokoban-reverse-{len(candidates)}",
                    actions=pushes,
                    seeds=(stable_hash(solved_state), depth, len(candidates)),
                    model_version="reverse_pull.sokoban.v1",
                )
                candidate = make_sokoban_candidate(layout, solved_state, predecessor, pushes, cost=depth)
                candidates.append((trace, candidate))
                queue.append(SokobanSearchNode(predecessor, pushes, depth))
                if len(candidates) >= max_candidates:
                    break
            if len(candidates) >= max_candidates:
                break
    return tuple(candidates), expansions


def search_sokoban_predecessor(
    layout: SokobanLayout,
    solved_state: SokobanState,
    max_depth: int = 3,
    max_candidates: int = 64,
) -> SokobanReverseReport:
    if not is_solved(layout, solved_state):
        raise ValueError("solved_state must have every box on a goal")
    traces, expansions = reverse_pull_traces(layout, solved_state, max_depth=max_depth, max_candidates=max_candidates)
    adapter = SokobanReverseAdapter()
    engine = TransactionEngine(adapter=adapter)
    predecessor = None
    pushes: tuple[PushAction, ...] = ()
    committed = False
    for trace, candidate in traces:
        outcome = engine.transact(solved_state, trace, candidate)
        if outcome.committed:
            predecessor = outcome.state
            pushes = tuple(_normalize_push(push) for push in candidate.payload["pushes"])
            committed = True
            break
    replay_rate = 0.0
    if engine.ledger.audit():
        try:
            replay_rate = 1.0 if engine.rollback_audit(solved_state) == solved_state else 0.0
            engine.replay_audit(solved_state)
        except Exception:
            replay_rate = 0.0
    baseline = _baseline_state_count(layout)
    return SokobanReverseReport(
        solved=committed,
        predecessor=predecessor,
        solved_state=solved_state,
        pushes=pushes,
        verifier_calls=engine.hard_verifier_calls,
        reverse_expansions=expansions,
        max_depth=max_depth,
        baseline_state_count=baseline,
        verifier_call_reduction=baseline / engine.hard_verifier_calls if engine.hard_verifier_calls else 0.0,
        invalid_commit_count=engine.invalid_commit_count,
        ledger_audit=engine.ledger.audit(),
        replay_rollback_rate=replay_rate,
    )


def _baseline_state_count(layout: SokobanLayout) -> int:
    floor = layout.height * layout.width - len(layout.walls)
    return max(0, floor * max(0, floor - 1))


def _normalize_state(state: SokobanState) -> SokobanState:
    return SokobanState(boxes=_normalize_boxes(state.boxes), player=tuple(state.player))  # type: ignore[arg-type]


def _normalize_boxes(boxes: Iterable[Position]) -> tuple[Position, ...]:
    normalized = tuple(sorted((int(row), int(col)) for row, col in boxes))
    if len(set(normalized)) != len(normalized):
        raise ValueError("duplicate sokoban boxes are not allowed")
    return normalized


def _normalize_push(push: PushAction) -> dict[str, Any]:
    direction = str(push["direction"])
    _direction_delta(direction)
    row, col = push["box"]
    return {"box": (int(row), int(col)), "direction": direction}


def _coerce_layout(value: Any) -> SokobanLayout:
    if isinstance(value, SokobanLayout):
        return value
    return SokobanLayout(
        height=int(value["height"]),
        width=int(value["width"]),
        walls=_normalize_boxes(value["walls"]),
        goals=_normalize_boxes(value["goals"]),
    )


def _coerce_state(value: Any) -> SokobanState:
    if isinstance(value, SokobanState):
        return _normalize_state(value)
    return SokobanState(boxes=_normalize_boxes(value["boxes"]), player=tuple(value["player"]))  # type: ignore[arg-type]


def _direction_delta(direction: str) -> Position:
    for name, delta in DIRECTIONS:
        if direction == name:
            return delta
    raise ValueError(f"unknown sokoban direction: {direction!r}")


def _state_key(state: SokobanState) -> tuple[tuple[Position, ...], Position]:
    normalized = _normalize_state(state)
    return normalized.boxes, normalized.player


def _is_open(layout: SokobanLayout, boxes: Iterable[Position], pos: Position) -> bool:
    return _in_bounds(layout, pos) and pos not in layout.wall_set and pos not in frozenset(boxes)


def _in_bounds(layout: SokobanLayout, pos: Position) -> bool:
    row, col = pos
    return 0 <= row < layout.height and 0 <= col < layout.width


def _add(pos: Position, delta: Position) -> Position:
    return pos[0] + delta[0], pos[1] + delta[1]


def _sub(pos: Position, delta: Position) -> Position:
    return pos[0] - delta[0], pos[1] - delta[1]
